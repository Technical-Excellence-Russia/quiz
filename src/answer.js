import React from "react"
import PropTypes from "prop-types"

export default function Answer(props) {

	const className = `row p-2 m-2 border rounded ${props.isOpen ? (props.answer.correct ? "border-success" : "border-danger") : "border-none"}`

	return (
		<div className={className}>
			<div className="col text-left form-check">
				<input id={props.index} type="checkbox" className="form-check-input"/>
				<label htmlFor={props.index} className="form-check-label">{props.answer.text}</label>
			</div>
		</div>
	)
}

Answer.propTypes = {
	index: PropTypes.string.isRequired,
	answer: PropTypes.object.isRequired,
	isOpen: PropTypes.bool.isRequired
}