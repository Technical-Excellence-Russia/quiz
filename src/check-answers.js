export default function CheckAnswers(answers) {
	return answers.reduce((answer, acc) => acc &= answer.correct && (answer.checked || false))
}