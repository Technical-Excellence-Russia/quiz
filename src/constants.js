export const INIT = "INIT"
export const GO_TO_GROUP = "GO_TO_GROUP"
export const NEXT_QUESTION = "NEXT_QUESTION"
export const SET_CHECK = "SET_CHECK"
export const TOGGLE_ANSWER = "SHOW_ANSWER"