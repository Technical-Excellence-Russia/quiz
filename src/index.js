import "bootstrap"
import "./index.scss"

import React from "react"
import ReactDOM from "react-dom"

import QuestionFactory from "../lib/question"
import ReducerFactory from "./reducer-factory"
import App from "./app"

const currentYear = new Date().getFullYear()
const reducer = ReducerFactory(window, new QuestionFactory())

ReactDOM.render(
	<div className="container">
		<div className="row">
			<div className="co mt-1 text-center">
				<h2 className="text-center">Technical Excellence Quiz</h2>
			</div>
		</div>
		<App Reducer={reducer}/>
		<div className="row m-3">
			<div className="col text-end">
				<p className="copyright">&copy; {currentYear} Technical Excellence Ru</p>
			</div>
			<div className="col-sm-1">&nbsp;</div>
		</div>
	</div>
	, document.getElementById("app"))
//import Question from "../lib/question"


//const question = new Question()

//const params = new URL(window.location.href).searchParams

/*new Vue({
    el: "#app",
    data: {
        currentYear: new Date().getFullYear().toString(),
        isAnswerOpen: false,
        answer: "",
        defaultQuestion: params.get("question"),
        question: {
            text: "",
            tags: [],
            answers: [],
            description: [],
        },
    },
    methods: {
        goToContents: function () {
            window.location.href = root;
        },
        openAnswer: function () {
            this.question.answers = this.question.answers.map((answer) => {
                answer.class = "row mt-2 border " + (answer.correct ? "border-success" : "border-danger");
                return answer;
            })

            const isCorrect = this.question.answers.filter((answer) => answer.correct !== (answer.selected || false)).length === 0;
            this.answer = isCorrect ? "Правильно! " : " Не верно ;( " + this.question.description;
            this.isAnswerOpen = !this.isAnswerOpen;
        },
        loadQuestion: function () {
            this.isAnswerOpen = false;

            let q = question.getQuestion(this.defaultQuestion) || {
                url: "",
                text: "Такого вопроса еще нет:(",
                answers: [],
                description: "Такого ответа еще нет:("
            };

            this.question = q;
            this.question.answers = this.question.answers.sort(() => Math.random() - 0.5);

            params.set("question", q.url);
            window.history.replaceState({}, "Technical Excellence Quiz", "?" + params.toString());

            this.question = q;
            this.defaultQuestion = null;
        }
    },
    mounted() {
        this.loadQuestion(null);
    }
});*/
