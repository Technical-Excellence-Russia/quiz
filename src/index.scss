@import "~bootstrap/scss/bootstrap.scss";

$font-color: #d3d3d3;
$border-color: #939393;
$background-color: #2b2b2b;

html, body, {
  color: $font-color;
  background: $background-color;
  font-family: "Lucida Console", monospace;
}

code {
  color: $font-color;
}

table th, td {
  border: 1px solid $border-color;
  padding: 1px 1px 1px 1px;
}

.border-none {
  border-color: $background-color !important;
}

.copyright {
  font-size: 8pt;
}

@media (min-width: 240px) {
  font-size: 12px;
}

@media (min-width: 400px) {
  font-size: 14px;
}

@media (min-width: 768px) {
  font-size: 16px;
}

@media (min-width: 1000px) {
  font-size: 18px;
}

