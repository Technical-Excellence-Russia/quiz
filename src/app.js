import React, {useReducer} from "react"
import PropTypes from "prop-types"

import Context from "./context"

import Result from "./components/result"
import Question from "./components/question"
import {INIT} from "./constants"


export default function App(props) {

	const [state, dispatch] = useReducer(
		props.Reducer,
		{},
		() => props.Reducer({}, {type: INIT})
	)

	return (
		<Context.Provider value={[state, dispatch]}>
			{!state.isResultOpen ?
				<Question state={state}/>
				:
				<Result result={{answerIndex: state.answerIndex, total: state.totalQuestions}}/>}
		</Context.Provider>
	)
}

App.propTypes = {
	Reducer: PropTypes.func.isRequired
}