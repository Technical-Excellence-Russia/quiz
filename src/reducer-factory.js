import {GO_TO_GROUP, INIT, NEXT_QUESTION, SET_CHECK, TOGGLE_ANSWER} from "./constants"
import CheckAnswers from "./check-answers"

export default function ReducerFactory(window, questionFactory) {
	const nextQuestion = () => questionFactory.getQuestion()

	const initialState = {
		questionIndex: 1,
		answerIndex: 0,
		totalQuestions: 5,
		isAnswerOpen: false,
		isResultOpen: false,
		question: nextQuestion()
	}

	return function Reducer(state, action) {
		switch (action.type) {
		case INIT: {
			return initialState
		}

		case SET_CHECK: {
			const answers = state.question.answers
			answers[action.index].checked = action.value
			return {
				...state,
				question: {
					...state.question,
					answers: answers
				}
			}
		}

		case TOGGLE_ANSWER: {
			return {
				...state,
				isAnswerOpen: !state.isAnswerOpen
			}
		}

		case NEXT_QUESTION: {
			state.answerIndex = state.answerIndex + (CheckAnswers(state.question.answers) ? 1 : 0)
			if (state.questionIndex === state.totalQuestions) {
				return {
					...state,
					isResultOpen: true
				}
			} else {
				return {
					...state,
					questionIndex: state.questionIndex + 1,
					isAnswerOpen: false,
					question: nextQuestion()
				}
			}
		}

		case GO_TO_GROUP: {
			window.open("https://t.me/technicalexcellenceru", "blank")
			return state
		}

		default: {
			return state
		}
		}
	}

}