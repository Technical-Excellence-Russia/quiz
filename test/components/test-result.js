import React from "react"
import renderer from "react-test-renderer"
import Context from "../../src/context"

import Result from "../../src/components/result"
import {GO_TO_GROUP, INIT} from "../../src/constants"

describe("Result should", () => {

	let component = null
	const state = {answerIndex: 5, totalQuestions: 5}
	const dispatch = (action) =>
		expect([INIT, GO_TO_GROUP].indexOf(action.type) >= 0).toBeTruthy()


	beforeAll(() => component = renderer.create(
		<Context.Provider value={[state, dispatch]}>
			<Result/>
		</Context.Provider>)
	)

	test("trigger go to group", () => {
		component.root.findByProps({className: "btn me-2 btn-primary"}).props.onClick()
		expect.assertions(1)
	})

	test("trigger close", () => {
		component.root.findByProps({className: "btn ms-2 btn-secondary"}).props.onClick()
		expect.assertions(1)
	})

	test("equal to snapshot", () => {
		expect(component.toJSON()).toMatchSnapshot()
	})

	afterAll(() => component.unmount())
})