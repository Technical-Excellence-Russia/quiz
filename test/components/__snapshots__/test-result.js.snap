// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result should equal to snapshot 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col m-2 p-2 border-top border-light"
    >
      Ты ответил правильно на 
      5
       из 
      5
      .
      <br />
      <br />
      А ещё у нас есть группа практиков;)
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col m-2 p-2 border-top border-light text-center"
    >
      <button
        className="btn me-2 btn-primary"
        onClick={[Function]}
        type="button"
      >
        В группу
      </button>
      <button
        className="btn ms-2 btn-secondary"
        onClick={[Function]}
        type="button"
      >
        Ещё раз
      </button>
    </div>
  </div>,
]
`;
